src/requirements/Allowlist/components/AllowlistFormInputs.tsx

```diff
@@ -49,102 +57,152 @@ export default function AllowlistFormInputs({
-        ?.filter((line) => !!line)
-        ?.map((line) => line.trim())
-
-      setValue(`${baseFieldPath}.data.addresses`, lines, {
-        shouldValidate: true,
-        shouldDirty: true,
-      })
-    }
-
-    fileReader.readAsText(file)
-  }
-
   const { isDragActive, fileRejections, getRootProps, getInputProps } = useDropzone({
     multiple: false,
     accept: { "text/*": [".csv", ".txt"] },
     onDrop: (accepted) => {
-      if (accepted.length > 0) parseFile(accepted[0])
+      if (accepted.length > 0) parseFile.onSubmit(accepted[0])
     },
+    maxSizeMb: MAX_IPFS_FILE_SIZE_MB,
   })
 
   const type = useWatch({ name: `${baseFieldPath}.type` })
+  const ipfsHash = useWatch({ name: `${baseFieldPath}.data.ipfsHash` })
+  const [uploadedLength, setUploadedLength] = useState<number>()
+
+  const { onUpload, progress } = usePinataWithProgress()
+
+  const parseFile = useSubmit(
+    (file: File) =>
+      new Promise<void>((resolve, reject) => {
+        setUploadedLength(null)
+        const fileReader = new FileReader()
+
+        fileReader.onload = () => {
+          const lines = fileReader.result
+            ?.toString()
+            ?.split("\n")
+            ?.filter((line) => !!line)
+            ?.map((line) => line.trim())
+
+          onUpload({
+            data: [JSON.stringify(lines)],
+            fileNames: ["allowlist.json"],
+          })
+            .then(({ IpfsHash }) => {
+              setUploadedLength(lines.length)
+              setValue(`${baseFieldPath}.data.ipfsHash`, IpfsHash)
+              setValue(`${baseFieldPath}.data.addresses`, [], {
+                shouldDirty: true,
+                shouldValidate: true,
+              })
+            })
+            .then(() => resolve())
+            .catch(() => {
+              reject(new Error("Failed to upload file to IPFS"))
+            })
+        }
+
+        fileReader.onerror = () => reject(new Error("Failed to read file"))
+
+        fileReader.readAsText(file)
+      })
+  )
 
   return (
     <Stack spacing={4} alignItems="start" {...getRootProps()}>
       <FormControl isInvalid={!!fileRejections?.[0]} textAlign="left">
         <FormLabel>Upload from file</FormLabel>
 
-        <Button as="label" leftIcon={<File />} h={10} maxW={56} cursor="pointer">
+        <Button
+          as="label"
+          leftIcon={<File />}
+          h={10}
+          maxW={56}
+          cursor="pointer"
+          isLoading={parseFile.isLoading}
+          loadingText={"Uploading"}
+          spinner={
+            <JoinStepIndicator
+              status={progress === 0 ? "LOADING" : "PROGRESS"}
+              progress={progress * 100}
+            />
+          }
+        >
           <input {...getInputProps()} hidden />
           <Text as="span" display="block" maxW={44} noOfLines={1}>
             {isDragActive ? "Drop the file here" : "Choose .txt/.csv"}
           </Text>
         </Button>
 
+        {typeof uploadedLength === "number" && (
```

Comment:
Consider using a `Collapse` here 😉

Reply:
Nice catch! Added in https://github.com/guildxyz/guild.xyz/pull/1149/commits/945a52fa9ab6b60b4de70e6711c3b37490ba8ed0, looks a lot smoother this way :monkayay: